### PR TITLE
Ignore expired certificates for OpenSSL

### PR DESCRIPTION
**Pull request by kouno**
I implemented the fix from @smuething in https://github.com/Homebrew/homebrew/issues/39870#issue-77627443.

Let me know if we can get this merged or if I messed up something on the way :smile:.

### CODE REVIEW

File: Library/Formula/openssl.rb [hunk 1]
```diff
@@ -121,6 +118,33 @@ def caveats; <<-EOS.undent
...
+
+  def get_system_certificates
+    keychains = %w[
+      /Library/Keychains/System.keychain
+      /System/Library/Keychains/SystemRootCertificates.keychain
+    ]
+
+    certs_list = `security find-certificate -a -p #{keychains.join(" ")}`
+    certs = certs_list.scan(
+      /#{Regexp.escape('-----BEGIN CERTIFICATE-----')}.*?#{Regexp.escape('-----END CERTIFICATE-----')}/m
```
Comment (jacknagel) [inline Library/Formula/openssl.rb]: There aren't any metacharacters here so we shouldn't need to escape the strings.

File: Library/Formula/openssl.rb [hunk 2]
```diff
@@ -121,6 +118,33 @@ def caveats; <<-EOS.undent
...
+    )
+
+    prune_expired_certificates(certs)
+  end
+
+  def prune_expired_certificates(certs)
+    certs.keep_if do |cert|
```
Comment (jacknagel) [inline Library/Formula/openssl.rb]: `certs.select`

File: Library/Formula/openssl.rb [hunk 3]
```diff
@@ -121,6 +118,33 @@ def caveats; <<-EOS.undent
...
+      IO.popen(%w[openssl x509 -inform pem -checkend 0 -noout], 'w') do |openssl_io|
```
Comment (jacknagel) [inline Library/Formula/openssl.rb]: Array form of IO.popen isn't available on Ruby 1.8

File: Library/Formula/openssl.rb (hunk 1 above)
Comment (kouno) [inline Library/Formula/openssl.rb]: I have no idea why I thought that `-` needed to be escaped. :)

**Comment by DomT4** [comment]
Why are we making the method `private` here?

File: Library/Formula/openssl.rb [hunk 4]
```diff
@@ -121,6 +118,33 @@ def caveats; <<-EOS.undent
...
+  def prune_expired_certificates(certs)
+    certs.select do |cert|
+      IO.popen('openssl x509 -inform pem -checkend 0 -noout', 'w') do |openssl_io|
+        openssl_io.write(cert)
+        openssl_io.close_write
+      end
+
+      $?.success?
+    end
+  end
```
Comment (mikemcquaid) [inline Library/Formula/openssl.rb]: Just inline these methods as they are only called once, thanks.

**Comment by kouno** [comment]
Updated after feedback.

**Comment by mikemcquaid** [comment]
@kouno you missed "Just inline these methods as they are only called once, thanks."

**Comment by kouno** [comment]
> @kouno you missed "Just inline these methods as they are only called once, thanks."

If you look at the diff, I removed `prune_expired_certificates`. Your comment is still showing because you commented on a line that I didn't modify.

**Comment by kouno** [comment]
Ah... sorry. I just got it. You meant both of them.

**Comment by DomT4** [comment]
Happy to pull this. It doesn't make a massive change to the cert list, a few hundred lines difference, but the duplication of expired/valid certificates in the Keychain is enough of a problem to merit action on our side and it's likely to help the situation.

Given Apple's long-term shift away from OpenSSL on OS X I'd be tempted to say the post-install usage of OpenSSL here should lean on the Homebrew OpenSSL, but I'll leave it for now and see how it goes.

We need to apply this to GnuTLS and LibreSSL as well. If you want to handle those with a new PR that'd be awesome. Thanks for your contribution to Homebrew, we appreciate it! :smiley_cat:.

**Pull request closed without merging**